src/groth16/prover/supraseal.rs

```diff
@@ -50,42 +46,48 @@ where
     for prover in &provers {
         assert_eq!(
             prover.a.len(),
-            n,
+            provers[0].a.len(),
             "only equaly sized circuits are supported"
         );
-        debug_assert_eq!(
-            a_aux_density_total,
-            prover.a_aux_density.get_total_density(),
-            "only identical circuits are supported"
-        );
-        debug_assert_eq!(
-            b_input_density_total,
-            prover.b_input_density.get_total_density(),
-            "only identical circuits are supported"
-        );
-        debug_assert_eq!(
-            b_aux_density_total,
-            prover.b_aux_density.get_total_density(),
-            "only identical circuits are supported"
-        );
     }
 
-    let mut input_assignments_ref = Vec::with_capacity(num_circuits);
-    let mut aux_assignments_ref = Vec::with_capacity(num_circuits);
-    for i in 0..num_circuits {
-        input_assignments_ref.push(input_assignments_no_repr[i].as_ptr());
-        aux_assignments_ref.push(aux_assignments_no_repr[i].as_ptr());
+    impl<Scalar> Into<supraseal_c2::Assignment<Scalar>> for &ProvingAssignment<Scalar>
```

Comment:
Somehow one of my comments didn't make it to GitHub, so let's try it again:

In Rust you usually implement `From` instead of `Into`. Quoting the [`From` documentation](https://doc.rust-lang.org/std/convert/trait.From.html):

> One should always prefer implementing `From` over `Into` because implementing From automatically provides one with an implementation of `Into` thanks to the blanket implementation in the standard library.

I would also move it to the top-level and not inside the function.

As I needed to try it you myself if it actually work, here's the code:

```rust
impl<Scalar> From<&ProvingAssignment<Scalar>> for supraseal_c2::Assignment<Scalar>
where
    Scalar: PrimeField,
{
    fn from(assignment: &ProvingAssignment<Scalar>) -> Self {
        assert_eq!(assignment.a.len(), assignment.b.len());
        assert_eq!(assignment.a.len(), assignment.c.len());

        Self {
            a_aux_density: assignment.a_aux_density.bv.as_raw_slice().as_ptr(),
            a_aux_bit_len: assignment.a_aux_density.bv.len(),
            a_aux_popcount: assignment.a_aux_density.get_total_density(),

            b_inp_density: assignment.b_input_density.bv.as_raw_slice().as_ptr(),
            b_inp_bit_len: assignment.b_input_density.bv.len(),
            b_inp_popcount: assignment.b_input_density.get_total_density(),

            b_aux_density: assignment.b_aux_density.bv.as_raw_slice().as_ptr(),
            b_aux_bit_len: assignment.b_aux_density.bv.len(),
            b_aux_popcount: assignment.b_aux_density.get_total_density(),

            a: assignment.a.as_ptr(),
            b: assignment.b.as_ptr(),
            c: assignment.c.as_ptr(),
            abc_size: assignment.a.len(),

            inp_assignment_data: assignment.input_assignment.as_ptr(),
            inp_assignment_size: assignment.input_assignment.len(),

            aux_assignment_data: assignment.aux_assignment.as_ptr(),
            aux_assignment_size: assignment.aux_assignment.len(),
        }
    }
}
```

Reply:
Fixed up. (Please excuse my distrust for compilers, expressed as preference to give the compiler exactly as much as required, it's kind of an occupational hazard:-)